qa/rpc-tests/finalsaplingroot.py

```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/env python2
+# Copyright (c) 2018 The Zcash developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+    connect_nodes_bi,
+    initialize_chain_clean,
+    start_nodes,
+    wait_and_assert_operationid_status,
+)
+
+from decimal import Decimal
+
+SAPLING_TREE_EMPTY_ROOT = "3e49b5f954aa9d3545bc6c37744661eea48d7c34e3000d82b7f0010c30f4c2fb"
+NULL_FIELD = "0000000000000000000000000000000000000000000000000000000000000000"
+
+# Verify block header field 'hashFinalSaplingRoot' (returned in rpc as 'finalsaplingroot')
+# is updated when Sapling transactions with outputs (commitments) are mined into a block.
+class FinalSaplingRootTest(BitcoinTestFramework):
+
+    def setup_chain(self):
+        print("Initializing test directory "+self.options.tmpdir)
+        initialize_chain_clean(self.options.tmpdir, 4)
+
+    def setup_network(self, split=False):
+        self.nodes = start_nodes(4, self.options.tmpdir, extra_args=[[
+            '-nuparams=5ba81b19:100', # Overwinter
+            '-nuparams=76b809bb:200', # Sapling
+            '-txindex'                # Avoid JSONRPC error: No information available about transaction
+            ]] * 4 )
+        connect_nodes_bi(self.nodes,0,1)
+        connect_nodes_bi(self.nodes,1,2)
+        connect_nodes_bi(self.nodes,0,2)
+        connect_nodes_bi(self.nodes,0,3)
+        self.is_network_split=False
+        self.sync_all()
+
+    def run_test(self):
+        # Activate Overwinter and Sapling
+        self.nodes[0].generate(200)
+        self.sync_all()
+
+        # Verfify genesis block contains null field for what is now called the final sapling root field.
+        blk = self.nodes[0].getblock("0")
+        assert_equal(blk["finalsaplingroot"], NULL_FIELD)
+
+        # Verify all generated blocks contain the empty root of the Sapling tree.
+        blockcount = self.nodes[0].getblockcount()
+        for height in xrange(1, blockcount + 1):
+            blk = self.nodes[0].getblock(str(height))
+            assert_equal(blk["finalsaplingroot"], SAPLING_TREE_EMPTY_ROOT)
+
+        # Node 0 shields some funds
+        taddr0 = self.nodes[0].getnewaddress()
+        saplingAddr0 = self.nodes[0].z_getnewaddress('sapling')
+        recipients = []
+        recipients.append({"address": saplingAddr0, "amount": Decimal('20')})
+        myopid = self.nodes[0].z_sendmany(taddr0, recipients, 1, 0)
+        mytxid = wait_and_assert_operationid_status(self.nodes[0], myopid)
+
+        self.sync_all()
+        self.nodes[0].generate(1)
+        self.sync_all()
+
+        # Verify the final Sapling root has changed
+        blk = self.nodes[0].getblock("201")
```

Comment:
Maybe index using the return from `getblockcount()`? That way (1) if more steps are added later, the test automatically adjusts, and (2) the reader is sure the test is referencing the most recent block.

Reply:
I considered it, but this test is small and it makes it easy to visually check what block we are checking.